src/openrct2/ride/TrackData.h

```diff
@@ -72,6 +72,13 @@ constexpr const dodgems_track_size DodgemsTrackSize(track_type_t type)
     return { 0, 0, 0, 0 };
 }
 
+class TrackElementFunction
```

Comment:
I'm against this type of polymorphism, this will introduce unwanted overhead. What you should use is a function prototype ```int32_t (*TrackEvaluatorFunction)(const uint16_t progress)``` which doesn't require vtables.

An even better way would be to have this type based such as ```template<TrackElemType TType> struct Evaluator { ... }``` which would also allow way better compiler optimizations also.

Reply:
The problem is that there are multiple places where we need to return a constant. I would need to declare a constant function for each value.

Reply:
Never mind, I think I know how to do it.

Reply:
> The problem is that there are multiple places where we need to return a constant. I would need to declare a constant function for each value.

You can also do this with the template system, something like ```Evaluator<Type, Constant>```

Reply:
Example https://godbolt.org/z/7ba5E9dbb